packages/deployer/subtasks/validate-namespaces.js

```diff
@@ -0,0 +1,49 @@
+const path = require('path');
+const rimraf = require('rimraf');
+const { subtask } = require('hardhat/config');
+const logger = require('@synthetixio/core-js/utils/logger');
+
+const {
+  getContractDataFromDebugFile,
+  getDebugDataFile,
+  getSlotAddress,
+  findDuplicateSlots,
+} = require('../internal/ast-helper');
+const { SUBTASK_VALIDATE_NAMESPACES } = require('../task-names');
+
+subtask(SUBTASK_VALIDATE_NAMESPACES).setAction(async (_, hre) => {
+  logger.subtitle('Validating namespaces');
+
+  const debugFiles = await _getDebugFiles(hre);
+  const slots = [];
+
+  debugFiles.forEach((file) => {
+    const { contractName, contractPath } = getContractDataFromDebugFile(file);
+    const debugDataFile = getDebugDataFile(file, contractPath);
+    const slotAddress = getSlotAddress(contractName, debugDataFile.output.sources);
+    slots.push({ contractName, slotAddress });
+  });
+
+  const duplicates = findDuplicateSlots(slots);
+
+  if (duplicates) {
+    const details = duplicates.map(
+      (d) => `  > ${d.slotAddress} found in storage contracts ${d.contracts}\n`
+    );
+
+    logger.error(`Duplicate Namespaces found!\n${details.join('')}`);
+    logger.info(`Deleting ${hre.deployer.deployment.file}`);
+    rimraf.sync(path.resolve(hre.config.paths.root, hre.deployer.deployment.file));
+    process.exit(0);
+  }
+
+  logger.checked('Namespaces are valid');
+});
+
+async function _getDebugFiles(hre) {
```

Comment:
I'm not sure if just looking at the expected folder where namespaces are declared is enough. Even if, as Mati says, we designate a "storage" folder in the plugin, developers may declare storage namespaces elsewhere, and they would be completely missed by this validation.

Reply:
Also, not just the location, but the time: We should probably not look at AST compiler data in hardhat, but in deployments. We should store AST trees, so that we can not only look at storage structure, but how it changes over time. This will be needed for determining if the struct within a namespace has been illegally reordered for example.